File: ga/glm4_result_2024_6/result/1968G2_glm/19_1968G2.c
use std::io::{self, BufRead};
use std::cmp::max;

const N: usize = 200000;
const N_: usize = 1 << 18; // N_ = pow2(ceil(log2(N)))

fn max(a: i32, b: i32) -> i32 {
    a.max(b)
}

fn build(aa: &mut Vec<i32>, n: usize) -> Vec<i32> {
    let mut st = vec![0; N_ * 2];
    let mut n_ = 1;
    while n_ < n {
        n_ <<= 1;
    }
    for i in 0..n {
        st[n_ + i] = aa[i];
    }
    for i in (1..n_).rev() {
        st[i] = max(st[i << 1], st[i << 1 | 1]);
    }
    st
}

fn query(st: &Vec<i32>, l: i32, a: i32) -> i32 {
    let mut l = l + N_;
    let mut r = N_ - 1;
    while l <= r {
        if l & 1 == 1 {
            if st[l] >= a {
                while l < N_ {
                    l = if st[l << 1] >= a { l << 1 } else { l << 1 | 1 };
                }
                return l - N_;
            }
            l += 1;
        }
        l >>= 1;
        r >>= 1;
    }
    N_
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();
    while t > 0 {
        let mut input = lines.next().unwrap().unwrap();
        let n: usize = input.parse().unwrap();
        let kl: usize = input.parse().unwrap();
        let kr: usize = input.parse().unwrap();
        let mut cc: String = input;
        let mut zz: Vec<i32> = vec![0; n];

        let mut i = 1;
        let mut l = 0;
        let mut r = 0;
        while i < n {
            if i + zz[i - l] < r {
                zz[i] = zz[i - l];
            } else {
                l = i;
                r = max(r, i);
                while r < n && &cc[r..r+l] == &cc[..l] {
                    r += 1;
                }
                zz[i] = r - l;
            }
            i += 1;
        }

        let mut st = build(&mut zz, n);
        let mut ans = vec![0; n + 1];

        for p in 1..=n {
            let mut k = 0;
            let mut i = 0;
            while i < n {
                i = query(&st, i as i32, p as i32);
                k += 1;
            }
            ans[k] = max(ans[k], p);
        }

        for k in (1..n).rev() {
            ans[k] = max(ans[k], ans[k + 1]);
        }

        for k in kl..=kr {
            print!("{} ", ans[k]);
        }
        println!();
        t -= 1;
    }
}

File: ga/glm4_result_2024_6/result/1968G2_cpp_glm/19_1968G2.rs
use std::io::{self, BufRead};
use std::cmp::Ordering;

const N: usize = 2e5 as usize + 3;

fn find(p: &mut Vec<usize>, k: usize) -> usize {
    if k > N || p[k] == k {
        k
    } else {
        p[k] = find(p, p[k]);
        p[k]
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let mut t = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
    while t > 0 {
        let mut n = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
        let mut x = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
        let mut y = lines.next().unwrap().unwrap().parse::<usize>().unwrap();
        let mut a = lines.next().unwrap().unwrap();
        let mut z = vec![0; N];
        let mut p = vec![0; N];
        let mut t = vec![0; N];
        let mut ans = vec![0; N];

        z[1] = n;
        for i in 2..=n {
            let mut j = 0;
            while j + z[j] > i {
                z[i] = std::cmp::min(z[i - j + 1], j + z[j] - i);
            }
            while a.chars().nth(i + z[i] - 1) == a.chars().nth(1 + z[i] - 1) {
                z[i] += 1;
            }
            if i + z[i] > j + z[j] {
                j = i;
            }
        }

        t[..n].sort_by(|&i, &j| z[i].cmp(&z[j]));
        for i in 1..=n {
            let mut k = 1;
            while k <= n && z[t[k]] < i {
                p[t[k]] = t[k] + 1;
                k += 1;
            }
            let mut c = 0;
            for j in 1..=n {
                c += 1;
                while find(&mut p, j + i) != j {
                    j = find(&mut p, j + i);
                }
            }
            ans[c] = i;
        }

        for i in (1..=n).rev() {
            ans[i] = std::cmp::max(ans[i], ans[i + 1]);
        }

        for i in x..=y {
            print!("{} ", ans[i]);
        }
        println!();

        t -= 1;
    }
}